HDDS-5755 EC: Refactor ECBlockOutputStreamEntry to accommodate all block group related ECBlockOuputStreams.

fapifta opened the pull request:
## What changes were proposed in this pull request?
Accomodate ECBlockOutputStreamEntry to handle the streams required for data and parity blocks.
Add changes to ECKeyoutputStream and ECBlockOutputStreamEntryPool, so that it uses entries as expected with this change.
This change will help to utilize preallocations properly, and simplifies the OMKeyLocationInfo assembly during commitKey, but does not touch how we handle the write and buffering of chuncks in ECKeyoutputStream.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-5755

## How was this patch tested?
Executed EC relevant tests, and new tests to check functionality.

fapifta commented:
@umamaheswararao you can take a look to the current state if you wish, I marked this as draft since there are still some junit tests I would like to add to it, and some things will require some polishing.
Here is what I am tracking:
- checkstyle for sure will fail, as there are some long lines remained
- ECBlockoutputstreamEntry#getTotalAckedDataLenght and ECBlockoutputStreamEntry#getWrittenDataLength does not consider parity, which I believed should fail some tests, but it did not, hence I would like to add some new tests before touching that (if you have any input on that feel free to let me know)
- Some method reordering and readability changes will come for sure.

umamaheswararao commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/TestECKeyOutputStream.java:
```diff
@@ -199,64 +191,6 @@ public void testCreateRatisKeyAndWithECBucketDefaults() throws Exception {
     }
   }
 
-
```
Why we removed this test?

fapifta replied:
This test is not feasible in the new code structure.
This test reaches out to ECBlockOutputStreams that were handling a single BlockOutputStreams earlier, and based on their Pipeline information it acquires a client from their XCeiverClientFactory, then checks if the acquired clients are distinct clients and none of them is equal to each other.
In the new code sturcture, the ECBlockOutputStreamEntry knows about the full EC Pipeline, and manages the data and parity streams together. Individual pipelines and streams are hidden outside.

The two new tests in the new ECBlockOutputStreamEntryTest class is entitled to test for the same, those tests are there to ensure that the Pipelines that are created by the ECBlockOutputStreamEntry#createSingleECBlockPipeline method are distinct in case the DataNodeDetails defines the same host but different port for EC replication, and I also added a test that ensures that if the port is also the same, then we reuse the client (the second test).

I believe this was important for tests where we need to ensure the tests work in a MiniCluster as well, where the DataNodeDetails differ in just the DataNode port, while in the Ratis replication case it was not really important as we do not differentiate blocks there while in EC we need to know the order of data blocks and parity, and we also need to communicate with all nodes and we can not share a client between them properly.

umamaheswararao commented on hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/ECBlockOutputStreamEntry.java:
```diff
@@ -17,53 +17,259 @@
+    // want to filter out parity here for example.
+//    return blockStreams()
+//        .mapToLong(BlockOutputStream::getTotalAckDataLength)
+//        .sum();
```
Why are we returning 0?

fapifta replied:
ECBlockOutputStream extends and extended BlockOutputStream before. Earlier ECBlockOutputStream did not override this method, so the method in BlockOutputStream was called. That method returned 0 hardwired.
RatisBlockOutputStream on the other hand maintains the acked data length via CommitWatcher, and returns that.

Our EC logic does not hold or check for any ack from the DNs via the grpc client, and does not maintain the acked legth so far, and it seems to be safer for now to just return 0 instead of calling the method of underlying streams, as we may not do this via those streams later on. I have added the naive implementation for discussion. I am unsure if the acked length should account for the parity block acks or for just the data block acks either.
And in any case, even if we sum the results, we will get 0 anyways, but it is easier to spot this way that the logic is not yet developed to calculate the acked length. Did I miss something?

umamaheswararao commented on hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/ECBlockOutputStreamEntry.java:
```diff
@@ -17,53 +17,259 @@
+    if (!isInitialized()) {
+      return 0;
+    }
+    return blockStreams()
```
Where are we skipping parity blocks length in blogGroup Len?

fapifta replied:
Next line limits the functional stream to just go over the data blocks. As the createStream logic creates all the streams for this blockGroup, it is safe to sum up the data length for the first replicationConfig.getData() streams in the array I believe. Do I misunderstand your question, or the contract of this method in the EC case?

umamaheswararao commented on hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/ECBlockOutputStreamEntry.java:
```diff
@@ -17,53 +17,259 @@
+
+  @Override
+  void createOutputStream() throws IOException {
+    Pipeline ecPipeline = getPipeline();
```
I think we don;t need to pass fill pipeline. WE can just create single node pipeline for that stream.

fapifta replied:
ECBlockOutputStreamEntry will manage the individual streams. In order for it to be able to create the streams and also to handle preallocations we need to have the full EC Pipeline here, as that will provide us all the necessary data to create the internal single replica pipelines (with that the BlockOutputStreams) to write the data to, and to report the key location info back to OM at the end of the write during commitKey.

I think there are 1 change is necessary here, instread of getPipeline() in the next line I should have use the ecPipeline to reference the full pipeline, making it obvious where the nodes are coming from, and to differentiate from the single replica pipelines created by the createSingleECBlockPipeline() method which provides them to the internal block output streams 8 lines below from here.

umamaheswararao commented on hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/ECBlockOutputStreamEntry.java:
```diff
@@ -17,53 +17,259 @@
+  boolean isClosed() {
+    if (!isInitialized()) {
+      return false;
+    }
```
What if some streams not initialized?
In the case when file hase 1MB, we will not intialize 2nd and 3rd streams as we don;t write anything there.

fapifta replied:
Initialization became a two stage process here, In the base BlockOutputStreamEntry class, there is the checkStream() method, which is the only place we are calling createOutputStream(). The createOutputStream method in one go initializes data+parity ECBlockOutputStreams, and places them into the blockOutputStreams array.
So that this class is initialized, if we have any non-null element in the blockOutputStreams array, therefore isInitialized checks if the first element of that array is non-null, and returns true if it is non-null.

The ECBlockOutputStreams themselves are cheap to create, and unless anything is written to one, they do not initiate any communication towards the DataNodes, nor do anything at close or flush.
With that I believe, we are fine initializing all the streams at once at createStream, and throw away some at the end without being used if we do not need it, this might be an optimization later, but I am unsure yet if we need that gain those microseconds at the time when we start writing after a block allocation, while we can spare the complexity it brings in if we instantiate the streams on demand.

umamaheswararao commented on hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java:
```diff
@@ -1071,7 +1071,7 @@ public OzoneOutputStream createMultipartKey(String volumeName,
     keyOutputStream.addPreallocateBlocks(
         openKey.getKeyInfo().getLatestVersionLocations(),
         openKey.getOpenVersion());
```
We may consider this changes to master?

umamaheswararao reviewed (commented):
Thanks @fapifta for good work. Changes looks good and approach is inline with what we discussed. 
I have dropped few early comments for your consideration to improve in the final draft. Thanks

fapifta commented on hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/KeyOutputStream.java:
```diff
@@ -27,7 +27,6 @@
 import java.util.stream.Collectors;
 
 import org.apache.hadoop.fs.FSExceptionMessages;
```
Added https://issues.apache.org/jira/browse/HDDS-5815 for this part, and will remove it from this PR.
The PR for HDDS-5815 is 2708.

fapifta commented on hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/BlockOutputStreamEntry.java:
```diff
@@ -36,7 +36,14 @@
 import com.google.common.annotations.VisibleForTesting;
 
 /**
```
I have consolidated the changes that can go into master under HDDS-5816. PR for that is up under #2709, I am preparing this PR to be based on those changes, so that we will need to wait until the changes are merged in master, and then merged to the base EC branch though.

fapifta commented on hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/BlockOutputStreamEntryPool.java:
```diff
@@ -156,10 +180,22 @@ void addKeyLocationInfo(OmKeyLocationInfo subKeyInfo) {
+    streamEntries.add(createStreamEntry(subKeyInfo));
   }
 
+  /**
```
They are added through #2709, I will be removing these from this PR, consequence is that this PR will be ready for review/merge after the master has the changes and those are merged to the base EC branch.

fapifta commented on hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/BlockOutputStreamEntryPool.java:
```diff
@@ -283,26 +338,21 @@ void commitKey(long offset) throws IOException {
-  }
-
+  /**
+   * Allocates a new block with OM, currently ECKeyOutputStream is an exclusive
```
I have corrected this, and also removed the EC related notes and descriptions in #2709

fapifta commented:
As we discussed today, it seems that the ECBlockInputStream change is unnecessary, and shed light on a bug in the mock cluster side in case of mock based tests, where the original pipeline we got from the allocateBlock, did not contain replicaIndicies in the pipeline, while in a real cluster it is there for EC.
I have removed the check from ECBlockInputStream, and added the indexes to the pipeline in the last commit.

@umamaheswararao as we discussed probably this is the last change necessary, but please kindly do the review and let me know if you see anything else with the current patch.
